### ISSUE TITLE
Git gutters *still* don't show on Uppercase filenames

### ISSUE THREAD

**Issue by mattrossman** [microsoft/vscode#103444]
- VSCode Version: 1.47.3 91899dcef7b8110878ea59626991a18c8a6a1b3e
- OS Version: MacOS 10.14.6

Steps to Reproduce:

1. Initialize a git repository
2. Add a capitalized filename, e.g. `Counter.jsx`
3. Commit the file
4. Make some changes in the file and save

The file will show as modified in the explorer, but the changes are not shown in the gutter. If you follow the same step but with a lowercase filename instead, the modifications will appear in the gutter. This is problematic for React projects where uppercase filenames are not uncommon.

This issue was already brought up in #48714 and closed by 375411e, but the issue still occurs. I'm opening this new issue since the old one is locked.

Does this issue occur when all extensions are disabled?: Yes

**Comment by joaomoreno** [comment]
So, the same steps with a `counter.jsx` works just fine?

Can you show me the entire contents of `Git: Show Git Output`, as you run through both scenarios: with uppercase and with lowercase?

**Comment by mattrossman** [comment]
In testing this I realize the steps I originally shared are incorrect, and the issue is more complex that I thought. It is not an issue with Uppercase files specifically, rather it has to do with changing case more generally. Specifically, it seems to happen when VSCode creates the original file, as I couldn't recreate it with a simple list of shell commands.

Here are a new list of steps that I have verified to recreate the issue, along with the output you requested:

1. Create a new project and open in VSCode
```bash
mkdir project && cd project
git init
code .
```

2. In the VSCode "Explorer" create a new file called "counter.jsx" with some content:

```jsx
// counter.jsx
let foo;
```

3. In the VSCode terminal, commit the file to git
```bash
git add .
git commit -m "Initial commit"
```
Note that if you make edits to `counter.jsx` at this point, they will show in the gutter.

4. Close the open file. In the VSCode terminal, rename the file (with git) and commit the change
```bash
git mv counter.jsx Counter.jsx
git commit -m "Rename"
```

5. Click "Refresh" in the VSCode "Explorer" to see the changed filename

Now when you edit `Counter.jsx`, the changes don't show in the gutter.

You can follow the same steps using an originally uppercase filename that is renamed to lowercase to achieve the same results.

Here is the content of `Git: Show Git Output` after following the steps described here:

<details>
<summary>Click to view contents</summary>
<pre>
Looking for git in: /usr/local/bin/git
Using git 2.27.0 from /usr/local/bin/git
> git rev-parse --show-toplevel
> git rev-parse --git-dir
Open repository: /Users/matt/src/temp/project
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
> git check-ignore -v -z --stdin
> git show --textconv :Other.jsx
> git ls-files --stage -- /Users/matt/src/temp/project/Other.jsx
> git config --global user.email
> git config --global user.name
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
> git check-ignore -v -z --stdin
> git show --textconv :counter.jsx
> git ls-files --stage -- /Users/matt/src/temp/project/counter.jsx
> git check-ignore -v -z --stdin
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
> git ls-files --stage -- /Users/matt/src/temp/project/counter.jsx
> git cat-file -s 7866c251d5ac36aaa11acd0c9a754538525cfe30
> git show --textconv :counter.jsx
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
> git rev-parse --symbolic-full-name master@{u}
fatal: no upstream configured for branch 'master'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
> git rev-parse --symbolic-full-name master@{u}
fatal: no upstream configured for branch 'master'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
> git show --textconv :counter.jsx
> git ls-files --stage -- /Users/matt/src/temp/project/counter.jsx
> git status -z -u
> git symbolic-ref --short HEAD
> git rev-parse master
> git rev-parse --symbolic-full-name master@{u}
fatal: no upstream configured for branch 'master'
> git for-each-ref --sort -committerdate --format %(refname) %(objectname)
> git remote --verbose
> git config --get commit.template
</pre>
</details>

**Comment by joaomoreno** [comment]
> Now when you edit `Counter.jsx`, the changes don't show in the gutter.

Can you try the same steps but rename the file to something else like `counter-2.jsx`. Do the changes appear in the gutter then?

**Comment by mattrossman** [comment]
Yes, if you rename it to something with the same case like `counter-new.jsx` then the changes appear in the gutter. Also, in that case there is no need to refresh the explorer to see the changed file name.

**Comment by PavanKu** [comment]
@mattrossman , @joaomoreno 
This issue looks related to [issue#100763](https://github.com/microsoft/vscode/issues/100763). Basically, vscode doesn't recognise the change in filename case.
try 
```
mv filename.js Filename.js // vscode File explorar and filename of tab won't reflect the change
mv filename.js fileName.js //vscode File explorar and filename of tab won't reflect the change
mv filename.js abc.js // vscode File explorar and filename of tab both recognise the change and update itself
```

That could be the reason why git gutter does not recognise the change.

**Comment by Oran** [comment]
Same Issue, but file name weren't changed
It fixes itself if you clone the repo again from remote

**Comment by joaomoreno** [comment]
> It fixes itself if you clone the repo again from remote

@mattrossman Can you confirm this?

**Comment by mattrossman** [comment]
I'm not sure what you want me to do, I'm not using a remote in my example so it's not clear when I should be pushing or cloning. @Oran seems to be having a different issue from me since they did not change filenames.

**Comment by joaomoreno** [comment]
@bpasero Here's another tough pickle for you. These are the steps: https://github.com/microsoft/vscode/issues/103444#issuecomment-666541057

What happens is: once the workbench knows about `counter.jsx`, it creates an editor model for it. When that file is renamed to `Counter.jsx`, and even after closing all text editors, upon opening the `Counter.jsx` text editor, its model still holds a reference to a URI which lists the lowercase variant of the file. Here's where SCM reads it and where this bug surfaces:

https://github.com/microsoft/vscode/blob/56adc7d930ad862c90f82964bb3f618d342ba177/src/vs/workbench/contrib/scm/browser/dirtydiffDecorator.ts#L1179-L1179

The bug surfaces since git is case sensitive, despite Windows being case insensitive.

More notes:

- If you leave the file open and reload, the bug remains: the lowercase URI still lingers on.
- You can only get rid of the bug by closing the editor and reloading. Upon opening it again, the feature works.

**Comment by bpasero** [comment]
This is a consequence of using `IUriIdentityService` for text models. Once this service is used to create a `URI` it will stay until you restart the app. We would need some way to invalidate cached `URI` maybe based on file events or other things to solve this properly.

**Comment by jrieken** [comment]
> We would need some way to invalidate cached URI maybe based on file events or other things to solve this properly.

Unsure - rename is happening on a case-insensitive file system and therefore the path (in the sense of identity) isn't changing. @joaomoreno can you make your cache aware of uri identify rules? You can use `IUriIdentityService#extUri` and `ResourceMap` with the "right" `toKey` function

**Comment by bpasero** [comment]
I thought the issue here was that `git` itself is case-sensitive, even on Windows and we rely on `git` to do its work. I am not sure we get a hook into `git`'s rules of how to compare paths.

**Comment by joaomoreno** [comment]
Exactly, @bpasero points to the right pickle. Git has a case sensitivity setting but this is a user configuration, we shouldn't really mess with it.

**Comment by jrieken** [comment]
Well, I guess the uri identity service won't help you then because it doesn't know what the actual representation is. If the git is so sensitive then I am afraid that you need to resolve the true casing before interacting with git.

**Comment by NeeEoo** [comment]
I was able to fix a file that had this problem by modifying the workspace state database.

By modifying `memento/workbench.parts.editor`, `memento/workbench.editors.files.textFileEditor` and `codelens/cache2` to have the correct case on the filename.

While I wouldn't recommend editing the workspace state database, I am just saying it to help the developers to fix this bug.

**Comment by PapaMarc** [comment]
relative to other closed threads that poke around with this issue from various angles, this seems like the place at this point.
Just wanted to share a related scenario so dev can appreciate relevance and get a sense of justification to address... 

Whereas IIS servers are case insensitive, Unix webservers typically are not. 
I have some ooold web pages, that were hosted on IIS servers (privately in homes initially, then on public MS/IIS based hosters) for years... recently on moving them to GitHub pages hosting, lots of stuff 'broke' (404s) in the process.

Reason being, the underlying servers on GitHub pages are not hosted via IIS on Windows which is case insensitive (same for NameCheap shared hosting requiring respect for explicit casing). In attempting to fixup the filenames (or foldernames) to match href's to those files, when using rename in the VSCode Explorer pane against the git repo... while the change appears in that VSCode pane, it (just changing an uppercase letter to lower for example in an otherwise unchanged filename) doesn't commit up to GitHub. 
[This thread also touches on some of the underlying issues](https://stackoverflow.com/questions/17683458/how-do-i-commit-case-sensitive-only-filename-changes-in-git ).

A good workaround more/less highlighted in that thread (they achieve this slightly differently with git command line as seen in detail in that thread)  which works well in the VSCode GUI is as follows: 
   -1. rename file Name.jpg to name1.jpg <the desired casing propagates because of the additional char being a recognized change)
   -2. commit removed file Name.jpg, and name1.jpg
   -3. rename file name1.jpg to name.jpg
   -4. commit change made in step3
(note: if you don't commit in #2, and just stage the change, and then stage the change in #3... then try to commit the 2 stages changes... you'll find the 2nd staging just nulls out your attempts to make any change and there's nothing to commit).

Thus, a vote of support for VSCode to recognize/support casing mods of an otherwise unchanged file or folder name be considered as a P3/'Nice to have' item, since non-explicit casing of folders or filenames used in URLs hosted on non-IIS servers results in 404 errors.

**Comment by moshest** [comment]
Please see a fix here: https://stackoverflow.com/a/11183844/518153

**Comment by Dionysusnu** [comment]
I'd just like to point out the bug also stretches to the built-in diff view. If you open that, select a range, right click and use "Stage selected ranges", it gets staged in git as a new file that has the original casing, with the additions from the selected range as its only content.

While that original-case file is staged, the gutters _do_ show up in the editor for the new-case file. Trying to stage something again in the diff view appends the additions from that range to the original-casing file.

It also creates an entry for the original-case file in the unstaged changes list. Viewing the diff of this shows it as the whole file being an addition, except for the range originally marked for staging.

Unstaging the original-case file in its entirety makes everything go back to the same state as before the first stage action, in which the file is shown only once in the changes window, and the gutters are not visible.

**Comment by o-l-a-v** [comment]
Does it act different if case sensitivity is enabled for the directory? Can be done on NTFS and ReFS / Dev Drive.

* About Dev Drive: <https://aka.ms/DevDrive>
* About case sensitivity: <https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity>
* Enable (CMD or PowerShell as admin): `fsutil.exe file setCaseSensitiveInfo <path> enable`
* Check: `fsutil.exe file queryCaseSensitiveInfo <path>`

**Comment by Seb105** [comment]
This issue still exists but I also found a relatively easy fix when dealing with large amounts of files (such as where you have changed the case of a directory), where the individual repair approach would be difficult.

1. Rename the directory containg all the errored files/dirs to something else
2. Commit this
3. Rename it back to what you want
4. Commit this

This fixed the issue for me.

**Comment by Seb105** [comment]
> @Seb105  if you squash those two commits will the problem reappear? Seems kinda clunky to need 2 commits to do just 1 rename.

Haven't tried but I don't think it would work. The important thing you're actually doing is untracking and re-tracking the file so git considers it an entirely new file. I think this would get squashed to a rename

**Comment by Peter-Schorn** [comment]
I am still experiencing this same issue. I originally created a file called `loadbwip.ts`. I later renamed it to `loadBwip.ts` (uppercase B). After doing this, any change I made to the file did not result in any git diff decorations in the editor gutter.

I fixed my issue by removing every reference to this file I could find in the `state.vscdb` sqlite file from my workspace storage.

(On macOS, the workspace storage is located at `~/Library/Application Support/Code/User/workspaceStorage`. Find the specific folder for your workspace by inspecting the `workspace.json` file in each folder, which should contain the folder path to your workspace.)

**Comment by thnee** [comment]
## Problem description

The gutter git diff works fine for all files, except one specific file, where it doesn't show at all.
Doing git diff with other git programs is working fine, problem only happens in VSCode.
Problem persists across `Git: Refresh` VSCode command, across VSCode restart, and across Windows reboot.

I have a file `pkg/myapp/src/lib/myfeature/Mycomponent.svelte`.

When clicking the file to open it, gutter git diff is not working, and VSCode git log shows this.

```
2026-08-24 11:49:42.809 [info] > git show --textconv :pkg/myapp/src/lib/myfeature/mycomponent.svelte [79ms]
2026-08-24 11:49:42.810 [warning] [GitFileSystemProvider][readFile] File not found - git:/c%3A/Users/Mattias/code/myproject/pkg/myapp/src/lib/myfeature/mycomponent.svelte.git?%7B%22path%22%3A%22c%3A%5C%5CUsers%5C%5CMattias%5C%5Ccode%5C%5Cmyproject%5C%5Cpkg%5C%5Cmyapp%5C%5Csrc%5C%5Clib%5C%5Cmyfeature%5C%5Cmycomponent.svelte%22%2C%22ref%22%3A%22%22%7D
```

Note the lowercase spelling of `mycomponent.svelte`.
The error itself makes sense, if running on a case-sensitive file system.
But this error should not be happening, because VSCode should not be using that filename.

The file was never committed in git as `mycomponent.svelte`, only as `Mycomponent.svelte`.
Also this file was never moved to a different location in the tree, and the parent folders were never renamed during this files life time.

I'm not sure what exactly caused this, perhaps I have accidentally renamed the file to lower case locally, and then renamed it back again?
In any case, this seems to be a cache-invalidation bug, local to VSCode itself.

## Solution

Ran this in Powershell.

```
Remove-Item -Recurse -Force "$env:APPDATA\Code\User\workspaceStorage\*"
```

Now, gutter git diff is working again, and VSCode git log shows this.

```
2026-08-24 12:10:04.581 [info] > git show --textconv :pkg/myapp/src/lib/myfeature/Mycomponent.svelte [111ms]
```

## VSCode info

```
Version: 1.134.0 (user setup)
Commit: 110a328ea54b42367b803ec53ee0bf52ef26b419
Date: 2026-08-18T18:24:44Z
Electron: 42.8.1
ElectronBuildId: 14906494
Chromium: 148.0.7778.280
Node.js: 24.18.1
V8: 14.8.178.38-electron.0
@github/copilot: 1.0.81-0
@github/copilot-sdk: 1.0.11
OS: 110a328ea54b42367b803ec53ee0bf52ef26b419
```
